### DISCUSSION TITLE
Searching for keywords in columns or json fields

### DISCUSSION THREAD

**Discussion by patronovski** [pocketbase/pocketbase#6088] [Q&A] [answered]
Hello, i just was wondering how we can do  an efficient search in pocketbase as its not supporting full-text search capabilities of SQLite.
The suggested approach is to use wildcards (%keyword%), but will this not decrease performance drastically, because no index is used if your search starts with a wildcard.

What i want is basically to search for customers by there name, i have one column for this (fullname). If a user types "John", or "Doe", i want in my search results to appear a customer with fullname = "John Doe".

Another question about search: lets say we have a table "order" and customer is embedded as json field in the orders table. How can we efficiently search for a order by typing the customer name? Is it possible to add an index to a json field?

I hope fulltext support will be added in the future in this amazing tool.

**Comment by ganigeorgiev** [comment]
I don't understand what do you mean by _"no index is used if your search starts with a wildcard"_.

If you define an index on the searched field then the query planner could potentially utilize it if it determines that it may resolve the query more efficiently with it but this will depend on what other fields your query rely on (including sort fields).

I'm not sure if this matter with LIKE expressions but if the index is not being used you could try defining your index with `COLLATE NOCASE` for the column so that it is case-insensitive.

It is not a full-text search, but when I need a individual words wildcard search, what I usually do is to split the search term if necessary and apply a `~` condition for each part, something like:

```js
const search = "John Doe"

await pb.collection("customers").getFullList({
    // generates: `fullname ~ "John" || fullname ~ "Doe"`
    filter: search.split(" ").filter(Boolean).map((part) => pb.filter("fullname ~ {:part}", { part })).join(" || "),
})
```

How performant the above would be will depend on your dataset but I find it to be sufficient for my own use case.

Full-text support for now is not planned for the reasons explained in https://github.com/pocketbase/pocketbase/issues/193.

**Reply by ganigeorgiev** [reply]
> Another question about search: lets say we have a table "order" and customer is embedded as json field in the orders table. How can we efficiently search for a order by typing the customer name? Is it possible to add an index to a json field?

Also I don't exactly understand what do you mean by this but you can also define index on expressions (in your case with `json_extract`) but I would advice to normalize your data and store it in a separate collection and not as `json` field.

**Reply by ganigeorgiev** [reply]
I forgot to mention that if you want to inspect the generated queries you can start the executable with `--dev`. You can also try to copy and adjust the generated query from the console and run it with the sqlite3 cli too directly against `pb_data/data.db` with `EXPLAIN QUERY PLAN yourQuery` and verify that your index is actually used.

**Comment by patronovski** [comment]
Hi, thanks for the quick response.

_"I don't understand what do you mean by "no index is used if your search starts with a wildcard"."_

What i mean by this is that a **contains** query will **not** make use of the index:
WHERE column LIKE 'something%     -> this **WILL** use the index
WHERE column LIKE '%something%     -> this will **NOT** use the index

_"Also I don't exactly understand what do you mean by this but you can also define index on expressions (in your case with json_extract) but I would advice to normalize your data and store it in a separate collection and not as json field."_

I have a customers table and i included the whole customer object as json field in the orders table. Why:
Lets say a customer John has some address and there is an order and invoice for this customer. If we change John's name or address in the customers table, we do not want this change in the order/invoice. Because the invoice was already created and closed. Thats why i just copy the customer info in the order to prevent this change to happen in the order/invoice table. 
In the order/invoice overview, i want to be able to search for invoices/order by customer name, but i am not sure if its a good idea to include the customer as json in de order, or maybe just copy the customer fields as columns in the order and create indices on these columns,

I will try your suggestion and see how it performs under different circumstances. 
Thanks again for your help and keep up the great work!

**Reply by ganigeorgiev** [reply, marked as answer]
> WHERE column LIKE '%something% -> this will NOT use the index

Hm, after testing locally and reading the [SQLite docs](https://www.sqlite.org/optoverview.html#the_like_optimization) in more details, there are several requirements for the LIKE optimizations to take place but indeed you are correct that not having a leading wildcard is one of those conditions.

Regarding the `json` field in the `orders` table - again, you should be able to define partial index/[index on expression](https://www.sqlite.org/expridx.html) with `json_extract` if you know the exact object fields but I guess it would be bound by the same LIKE requirements as the regular columns so I'm not sure how useful that would be.

Based on my own production usage in [app.presentator.io](https://app.presentator.io/) and testing from the Dashboard against one of my biggest collection with a little over 350k records, the response time from 5 consequent searches is between `0.61801ms - 311.367427ms` (_and keep in mind that the Dashboard wild card search term is applied against each 10 column of the collection, aka. `id ~ "app" || title ~ "app" || created ~ "app" ...`_). The number of the found records in my case is 8714. The json object size of a single random picked record is 378B.

For my own specific use case I'm OK with the above performance, but for you that may not be the case.

My recommendation is to try seeding a test dummy database to simulate your expected load and make your decision based on that (you can find an example command how to import json file in https://github.com/pocketbase/pocketbase/discussions/6040#discussioncomment-11490098).

**Reply by patronovski** [reply]
Thanks Gani, i will experiment with it. 
Btw: the the fact that the index is not used when using a wildcard as prefix (i think) is the case with all major SQL databases, so not only with SQLite. Thats why fulltext is a better option for textual searches if you want to do something like "contains". Leading wildcards will always do a full table scan. A "startsWith" (wildcard as suffix) however will always use the index, so no problem with that. 

Thanks again and we love your work on this project!
